File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <stack>
#include <string>

using namespace std;

int main() {

	int t;
	cin >> t;

	while (t--) {
		
		int cnt = 0;
		stack<char> s;
		string str;
		cin >> str;

		for (char ch : str) {



			if (ch == '(') {
				
				cnt++;
				s.push(ch);
			}

			else {

				cnt--;

				if (s.empty())
					break;

				s.pop();
			}
		}

		if (cnt == 0) {

			cout << "YES" << endl;
		}

		else {
			
			cout << "NO" << endl;
		}
	}


	return 0;
}

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.*;
import java.util.*;
/*
if (push -> ")" 일때, peek()가 "("면? pop);
*/
public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        for (int i = 0; i < n; i++) {
            Stack<Character> sta = new Stack<>();
            String str = br.readLine();

            for (int j = 0; j < str.length(); j++) {
                char ch = str.charAt(j);

                if (ch == ')')
                    if (!sta.isEmpty() && sta.peek() == '(')
                        sta.pop();
                    else
                        sta.push(ch);
                else
                    sta.push(ch);
            }

            if (sta.isEmpty())
                System.out.println("YES");
            else
                System.out.println("NO");
        }
    }
}
